File: examples/python/example-actor-definition.py
# SPDX-License-Identifier: BSL-1.0

import time
import traeger


class Account(traeger.Actor):
    def __init__(self, initial_funds):
        self.funds = initial_funds

    def deposit(self, amount: float) -> float:
        # simulate a costly operation
        time.sleep(0.200)
        if amount <= 0:
            raise RuntimeError("invalid amount")
        self.funds += amount
        return self.funds

    def debit(self, amount: float) -> float:
        # simulate a costly operation
        time.sleep(0.200)
        if amount <= 0:
            raise RuntimeError("invalid amount")
        if self.funds < amount:
            raise RuntimeError("not enough funds")
        self.funds -= amount
        return self.funds

    def balance(self) -> float:
        # simulate a costly operation
        time.sleep(0.200)
        return self.funds

def make_account_actor(initial_funds: float) -> traeger.Actor:
    account_actor = traeger.make_actor(Account, initial_funds)
    account_actor.define_writer("deposit", Account.deposit)
    account_actor.define_writer("debit", Account.debit)
    account_actor.define_reader("balance", Account.balance)
    return account_actor

File: examples/example-actor-definition.cpp
// SPDX-License-Identifier: BSL-1.0

#include <thread>
#include <chrono>

#include <traeger/actor/Actor.hpp>

class Account
{
    traeger::Float funds_;

public:
    explicit Account(const traeger::Float initial_funds) noexcept
        : funds_(initial_funds)
    {
    }

    // this method is non-const i.e. it modifies the instance
    auto deposit(const traeger::Float amount) -> traeger::Float
    {
        // simulate a costly operation
        std::this_thread::sleep_for(std::chrono::milliseconds(200));
        if (amount <= 0)
        {
            throw std::runtime_error("invalid amount");
        }
        funds_ += amount;
        return funds_;
    }

    // this method is non-const i.e. it modifies the instance
    auto debit(const traeger::Float amount) -> traeger::Float
    {
        // simulate a costly operation
        std::this_thread::sleep_for(std::chrono::milliseconds(200));
        if (amount <= 0)
        {
            throw std::runtime_error("invalid amount");
        }
        if (funds_ < amount)
        {
            throw std::runtime_error("not enough funds");
        }
        funds_ -= amount;
        return funds_;
    }

    // this method is const i.e. it does not modify the instance
    auto balance() const noexcept -> traeger::Float
    {
        // simulate a costly operation
        std::this_thread::sleep_for(std::chrono::milliseconds(200));
        return funds_;
    }
};

// this function acts as the factory for the Account actor
traeger::Actor make_account_actor(traeger::Float initial_funds)
{
    auto account_actor = traeger::make_actor<Account>(initial_funds);
    account_actor.define("deposit", &Account::deposit);
    account_actor.define("debit", &Account::debit);
    account_actor.define("balance", &Account::balance);
    return static_cast<traeger::Actor>(std::move(account_actor));
}
